tsdb/index/inmem/meta.go

```diff
@@ -1367,35 +1339,42 @@ func (t *TagKeyValue) Range(f func(tagValue string, a SeriesIDs) bool) {
 
 // RangeAll calls f sequentially on each key and value. A call to RangeAll on a
 // nil TagKeyValue is a no-op.
-func (t *TagKeyValue) RangeAll(f func(k string, a SeriesIDs)) {
-	t.Range(func(k string, a SeriesIDs) bool {
+func (t *tagKeyValue) RangeAll(f func(k string, a seriesIDs)) {
+	t.Range(func(k string, a seriesIDs) bool {
 		f(k, a)
 		return true
 	})
 }
 
 // Store stores ids under the value key.
-func (t *TagKeyValue) Store(value string, ids SeriesIDs) {
+func (t *tagKeyValue) Store(value string, ids seriesIDs) {
 	t.mu.Lock()
-	defer t.mu.Unlock()
 	t.valueIDs[value] = ids
+	t.mu.Unlock()
+}
+
+// StoreByte stores ids under the value key.
```

Comment:
It's not clear to me why we need this new method.

Reply:
To complement the `LoadByte` as there is a drive to move to `[]byte` keys vs `string` keys. This change also removes a call to `runtime.slicebytetostring` at the call site in `measurement#AddSeries`.

Reply:
I get the push to `[]byte`; it's something that's been slowly chipped away at over time but seems never ending 😄 

> The cast in StoreByte is for the map key, so we benefit from the compiler optimization that it skips the alloc.

That was kind of the underlying point of my confusion I guess.. We can't benefit from that particular compiler optimisation here because we need to alloc in order to guarantee that the map key is stable don't we? I'm pretty sure that particular optimisation would only work on the read-side wouldn't it?

With `Store(value string, ids seriesIDs)` I would assume there would be a `runtime.slicebytetostring` on the `string(value)` at the call site but then there would be no further allocations would there? Or maybe there would be another allocation when we did `t.valueIDs[value] = ids` in `Store`?

Reply:
Indeed you are correct. Whilst map's won't allocate if a space exists to store a new key in a bucket, the assignment does indeed cause an allocation every time. It seems the compiler is not smart enough to delay the allocation _if_ the key does not exist.

However, that leads to another issue with this logic anyhow. The `*measurement#AddSeries` method calls `tagKeyValue#LoadByte` to fetch the seriesIDs slice and later overwrites with the `StoreByte` method. This is a race condition, as if two concurrent processes mutated the slice, one would overwrite the other on the `StoreByte` call. We know this won't happen, as it is only mutated via `*measurement#AddSeries` that takes an exclusive lock on the `measurement`. We should just remove all this complicated locking logic :)

Reply:
If you're referring to the lock on `TagKeyValue`, it was added specifically to fix an existing race (I forget where).